src/gennylib/src/PoolManager.cpp

```diff
@@ -40,28 +41,29 @@ auto createPool(const std::string& mongoUri,
 
 }  // namespace
 
+}  // namespace genny
+
 
 mongocxx::pool::entry genny::PoolManager::client(const std::string& name,
-                                                 unsigned long instance,
+                                                 size_t instance,
                                                  genny::WorkloadContext& context) {
     // Only one thread can access pools.operator[] at a time...
-    this->_poolsGet.lock();
+    std::unique_lock<std::mutex> getLock{this->_poolsGet};
     LockAndPools& lap = this->_pools[name];
     // ...but no need to keep the lock open past this.
     // Two threads trying access client("foo",0) at the same
     // time will subsequently block on the unique_lock.
-    this->_poolsGet.unlock();
+    getLock.unlock();
 
-    // only one thread can access the vector of pools at once
-    std::unique_lock lock{lap.first};
+    // only one thread can access the map of pools at once
+    std::unique_lock<std::mutex> lock{lap.first};
 
     Pools& pools = lap.second;
 
-    while (pools.empty() || pools.size() - 1 < instance) {
-        pools.push_back(createPool(this->_mongoUri, this->_apmCallback, context));
+    if (pools.find(instance) == pools.end()) {
```

Comment:
I'd still recommend writing this using `operator[]` so you aren't doing two lookups into the map. It'll default construct a `std::unique_ptr<mongocxx::pool>` for you and hence be nullptr.

```cpp
auto& pool = pools[instance];
if (!pool) {
    pool = createPool(this->_mongoUri, this->_apmCallback, context);
}
```

Reply:
Good idea - done.